src/main/java/graphql/schema/visitor/GraphQLSmartTypeVisitor.java

```diff
@@ -0,0 +1,52 @@
+package graphql.schema.visitor;
+
+import graphql.PublicSpi;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.schema.GraphQLFieldsContainer;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLTypeVisitor;
+import graphql.util.TraversalControl;
+
+/**
+ * I called it smart because I want to offer more "smarts" above GraphQLTypeVisitor and its "context" uber object
+ * <p>
+ * You would use it by doing `new GraphQLSmartTypeVisitor() { ...}.toTypeVisitor()`
+ */
+@PublicSpi
+public interface GraphQLSmartTypeVisitor {
+
+    default TraversalControl visitGraphQLObjectType(GraphQLObjectType objectType, SmartTypeVisitorEnvironment environment) {
+        return TraversalControl.CONTINUE;
+    }
+
+    /**
+     * This is a class specific for visiting {@link GraphQLFieldDefinition}s
+     */
+    interface GraphQLFieldDefinitionVisitorEnvironment extends SmartTypeVisitorEnvironment {
+
+        GraphQLFieldsContainer getFieldsContainer();
+
+    }
+
+    /**
+     * Each method has a specific env say or some do - TBD
+     *
+     * @param fieldDefinition the field
+     * @param environment     the specific env
+     *
+     * @return control
+     */
+    default TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, GraphQLFieldDefinitionVisitorEnvironment environment) {
+        return TraversalControl.CONTINUE;
+    }
+
+
```

Comment:
Clearly I need to add all the other visi methods - eg arguments, directives, unions and so on